packages/moonstone/VirtualList/VirtualListBase.js

```diff
@@ -466,82 +527,129 @@ const VirtualListBaseFactory = (type) => {
+								animate = (wrap === true);
+								isWrapped = true;
+							}
+						}
 
-				for (let i = currentIndex - 2; i >= 0; i--) {
-					if (!isItemDisabled(i)) {
-						nextIndex = i;
-						break;
+						// If there is no item which could get focus forward,
+						// we need to set restriction option to `self-first`.
+						if (nextIndex === -1) {
+							this.setRestrict(false);
+						}
+					} else {
+						nextIndex = this.findNearestSpottableItemInExtent(currentIndex, spottableExtent);
 					}
 				}
-
-				// If there is no item which could get focus backward,
-				// we need to set restriction option to `self-first`.
-				if (nextIndex === -1) {
-					this.setRestrict(false);
-				}
-			} else {
-				return false;
 			}
 
-			if (nextIndex !== -1 && (firstIndex > nextIndex || nextIndex >= firstIndex + numOfItems)) {
-				// When changing from "pointer" mode to "5way key" mode,
-				// a pointer is hidden and a last focused item get focused after 30ms.
-				// To make sure the item to be blurred after that, we used 50ms.
-				setTimeout(() => {
-					target.blur();
-				}, 50);
+			if (nextIndex !== -1) {
+				if (firstIndex <= nextIndex && nextIndex < firstIndex + numOfItems) {
+					this.focusOnItem(nextIndex);
+				} else {
+					this.nodeIndexToBeFocused = this.lastFocusedIndex = nextIndex;
+
+					if (!Spotlight.isPaused()) {
+						Spotlight.pause();
+					}
 
-				this.nodeIndexToBeFocused = this.lastFocusedIndex = nextIndex;
+					if (isWrapped) {
+						// In case of 'wrapping-around',
+						// we need to blur the current focus immediately
+						// since it can be a very long scroll (from one edge to the other edge)
+						// and definitely it's not a case of changing "pointer" mode to "5way key" mode.
+						target.blur();
+					} else {
+						// When changing from "pointer" mode to "5way key" mode,
+						// a pointer is hidden and a last focused item get focused after 30ms.
+						// To make sure the item to be blurred after that, we used 50ms.
+						setTimeout(() => {
+							target.blur();
+						}, 50);
+					}
 
-				if (!Spotlight.isPaused()) {
-					Spotlight.pause();
+					cbScrollTo({
+						index: nextIndex,
+						stickTo: isForward ? 'end' : 'start',
+						animate
+					});
 				}
 
-				cbScrollTo({
-					index: nextIndex,
-					stickTo: isForward ? 'end' : 'start'
-				});
 				return true;
 			}
 
 			return false;
 		}
 
 		onKeyDown = (ev) => {
-			const {keyCode, target} = ev;
+			const {keyCode, repeat, target} = ev;
 
 			this.isScrolledBy5way = false;
 			if (getDirection(keyCode)) {
 				ev.preventDefault();
 				this.setSpotlightContainerRestrict(keyCode, target);
-				this.isScrolledBy5way = this.jumpToSpottableItem(keyCode, target);
+				this.isScrolledBy5way = this.jumpToSpottableItem(keyCode, repeat, target);
+				if (this.shouldPreventScrollByFocus()) {
+					ev.stopPropagation();
+				}
 			}
-			forward('onKeyDown', ev, this.props);
```

Comment:
If removing this code, page up/down key handling code won't be called. Please check it again.